lib/Geo/GeoJson.cpp

```diff
@@ -39,616 +41,765 @@
-    return TRI_ERROR_BAD_PARAMETER;
+  removeAdjacentDuplicates(vertices);
+  if (Validation && ADB_UNLIKELY(vertices.size() < 2)) {
+    return {TRI_ERROR_BAD_PARAMETER,
+            "Invalid LineString,"
+            " adjacent vertices must not be identical or antipodal."};
   }
+  return {};
+}
 
-  std::vector<S2Point> vertices;
-  Result res = ::parsePoints(vpack, true, vertices);
-  if (res.ok()) {
-    if (vertices.empty()) {
-      return Result(TRI_ERROR_BAD_PARAMETER,
-                    "Invalid MultiPoint, must contain at least one point.");
+template<bool Validation>
+Result parseLinesImpl(velocypack::Slice vpack, std::vector<S2Polyline>& lines,
+                      std::vector<S2Point>& vertices) {
+  TRI_ASSERT(vpack.isArray());
+  velocypack::ArrayIterator it{vpack};
+  lines.clear();
+  lines.reserve(it.size());
+  for (; it.valid(); it.next()) {
+    auto slice = *it;
+    if (Validation && ADB_UNLIKELY(!slice.isArray())) {
+      return {TRI_ERROR_BAD_PARAMETER, "Missing coordinates."};
     }
-    region.reset(new S2MultiPointRegion(&vertices),
-                 ShapeContainer::Type::S2_MULTIPOINT);
+    auto r = parseLineImpl<Validation>(slice, vertices);
+    if (Validation && ADB_UNLIKELY(!r.ok())) {
+      return r;
+    }
+    // TODO(MBkkt) single unnecessary copy here, needs to patch S2!
+    lines.emplace_back(vertices, S2Debug::DISABLE);
+    TRI_ASSERT(lines.back().IsValid());
+  }
+  if (Validation && ADB_UNLIKELY(lines.empty())) {
+    return {
+        TRI_ERROR_BAD_PARAMETER,
+        "Invalid MultiLinestring, it must contains at least one Linestring."};
   }
-  return res;
+  return {};
 }
 
-/// https://tools.ietf.org/html/rfc7946#section-3.1.6
-/// First Loop represent the outer bound, subsequent loops must be holes
-/// { "type": "Polygon",
-///   "coordinates": [
-///    [ [100.0, 0.0], [101.0, 0.0], [101.0, 1.0], [100.0, 1.0], [100.0, 0.0] ],
-///    [ [100.2, 0.2], [100.8, 0.2], [100.8, 0.8], [100.2, 0.8], [100.2, 0.2] ]
-///   ]
-/// }
-Result parsePolygon(VPackSlice const& vpack, ShapeContainer& region,
-                    bool legacy) {
-  if (Type::POLYGON != type(vpack)) {
-    return {TRI_ERROR_BAD_PARAMETER, "requires type: 'Polygon'"};
+template<bool Validation>
+Result makeLoopValid(std::vector<S2Point>& vertices) noexcept(!Validation) {
+  if (Validation && ADB_UNLIKELY(vertices.size() < 4)) {
+    return {TRI_ERROR_BAD_PARAMETER,
+            "Invalid GeoJson Loop, must have at least 4 vertices"};
   }
-  TRI_ASSERT(Type::POLYGON == type(vpack));
-
-  VPackSlice coordinates = vpack;
-  if (vpack.isObject()) {
-    coordinates = vpack.get(Fields::kCoordinates);
+  // S2Loop doesn't like duplicates
+  removeAdjacentDuplicates(vertices);
+  if (Validation && ADB_UNLIKELY(vertices.front() != vertices.back())) {
+    return {TRI_ERROR_BAD_PARAMETER, "Loop not closed"};
+  }
+  // S2Loop automatically add last edge
+  if (ADB_LIKELY(vertices.size() != 1)) {
```

Comment:
0 is also "!=1" But popping on empty vector is UB
```suggestion
  if (ADB_LIKELY(vertices.size() > 1)) {
```